extensions/rates/functions/src/index.ts

```diff
@@ -21,35 +23,25 @@ logs.init(config);
 
 export const getRates = functions.handler.firestore.document.onWrite(
   async (change: Change<DocumentSnapshot>): Promise<void> => {
-    if (!change.after.exists) return; // The document is being deleted
-
-    let data = change.after.data() as InputPayload;
-
-    // Support situations where the keys may be snake_cased
-    data = camelizeKeys(data) as InputPayload;
+    if (hasInputChanged(change, castParams)) {
+      let data = change.after.data() as InputPayload;
 
-    // if (is.not.empty(data[config.shipmentKey])) return; // Rates has been completed
-    if (hasRatesData(data)) return;
```

Comment:
This would make it impossible to re-calculate rates when you change something about the shipment. We only need to block execution of the purchase-label if there's already a label there, so we don't buy multiple labels. Instead, we need to check to see if the input data we care about has changed between the previous update and this one.